File: solutions/python/sliding-window/get-equal-substrings-within-budget.py
# LeetCode 1208 — Get Equal Substrings Within Budget (Medium)
# Category: Sliding Window · Approach: Sliding Window
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/get-equal-substrings-within-budget/

def equal_substring(s: str, t: str, max_cost: int) -> int:
    left = cost = best = 0
    for right in range(len(s)):
        cost += abs(ord(s[right]) - ord(t[right]))
        while cost > max_cost:
            cost -= abs(ord(s[left]) - ord(t[left]))
            left += 1
        best = max(best, right - left + 1)
    return best

File: solutions/typescript/sliding-window/get-equal-substrings-within-budget.ts
// LeetCode 1208 — Get Equal Substrings Within Budget (Medium)
// Category: Sliding Window · Approach: Sliding Window
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/get-equal-substrings-within-budget/

function equalSubstring(s: string, t: string, maxCost: number): number {
  let left = 0, cost = 0, best = 0;
  for (let right = 0; right < s.length; right++) {
    cost += Math.abs(s.charCodeAt(right) - t.charCodeAt(right));
    while (cost > maxCost) {
      cost -= Math.abs(s.charCodeAt(left) - t.charCodeAt(left));
      left++;
    }
    best = Math.max(best, right - left + 1);
  }
  return best;
}
